src/Plugin/EntityBrowser/Widget/Webdam.php

```diff
@@ -0,0 +1,286 @@
+        '#type' => 'button',
+        '#value' => $folder_name,
+        '#name' => 'webdam_folder',
+        '#webdam_folder_id' => $folder_id,
+        '#webdam_parent_folder_id' => $folder_name,
+      ];
+    }
+
+    //Add container for assets (and folder buttons)
+    $form['asset-container'] = [
+      '#type' => 'container',
+    ];
+    //Add folder buttons to form
+    foreach ($folders as $folder){
+      $form['asset-container'][$folder->id] = [
+        '#type' => 'button',
+        '#value' => $folder->name,
+        '#name' => 'webdam_folder',
+        '#webdam_folder_id' => $folder->id,
+        '#webdam_parent_folder_id' => $current_folder->parent,
+      ];
+    }
+    //Assets are rendered as #options for a checkboxes element.  Start with an empty array.
+    $assets = [];
+    //Add to the assets array
+    foreach ($folder_items as $folder_item) {
+      if($folder_item->type == 'asset'){
+        //wrap asset name in a paragraph tag
+        $assets[$folder_item->id] = '<p class="webdam-asset-fielname">'.$folder_item->name.'</p>';
+        //If the asset has a thumbnail it should be rendered
+        if(!empty($folder_item->thumbnailurls)){
+          //Multiple thumbnail sizes are available.  Using the 220 width version for now
+          $assets[$folder_item->id] .= '<img class="webdam-asset-thumbnail" src="'.$folder_item->thumbnailurls[2]->url.'" alt="'.$folder_item->name.'" />';
+        }
+      }
+    }
+    //If the assets array is not empty then add the assets to the form as checkboxes
+    if(!empty($assets)){
+      $form['asset-container']['assets'] = [
+        '#type' => 'checkboxes',
+        '#title' => $this->t('Choose one or more assets'),
+        '#title_display' => 'invisible',
+        '#options' => $assets,
+        // Multiple assets will only be accepted if the source field allows more than one value.
+        '#multiple' => $field_cardinality != 1 && $this->configuration['multiple'],
+      ];
+    }
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function prepareEntities(array $form, FormStateInterface $form_state) {
+    foreach ($form_state->getValue(['upload'], []) as $aid) {
+      if ($aid !== 0) {
+        //TODO: validate form selection
+      }
+    }
+    return [];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submit(array &$element, array &$form, FormStateInterface $form_state) {
+    $assets = [];
+    if (!empty($form_state->getTriggeringElement()['#eb_widget_main_submit'])) {
+      foreach ($form_state->getValue(['assets'], []) as $aid) {
+        if ($aid !== 0) {
+          $webdam_asset = $this->webdam->getAsset($aid);
+          ksm($webdam_asset);
+          $media_asset = Media::create([
+            'bundle' => 'webdam',
+            'uid' => '1',
+            'langcode' => 'en',
+            'status' => Media::PUBLISHED,
+            'name' => $webdam_asset->name,
+            'field_asset_id' => $aid,
```

Comment:
The field where the asset ID goes should probably be configurable as well since we don't have any way of guaranteeing this field is present.

Reply:
Should we add it to the WebdamAsset media entity?  It seems like it should be a required field for this type of entity.  I see a 'type_id' field in the WebdamAsset.  Is that supposed to be the asset id?

Reply:
We cannot assume that a particular field exists, even if we create it in lightning_webdam. This module should be able to be used in a standalone way, so even if it's just a text field for now where you'd enter "field_asset_id" or whatever, that'd be fine. Ideally, when they select a media bundle, there would be a second dropdown that lists all of the integer fields on that media bundle, but that's a bit much for a first pass.

The type_id is a different thing. That's the type of asset on Webdam's side. The asset id can be pulled from any integer field.

Reply:
Resolved